### PR TITLE
master: extract common logic to a function

### LINKED ISSUE #811: We have so many types of protobuf message!
## Feature Request

**Is your feature request related to a problem? Please describe:**

I'm frustrated when adding automatic retry to leader-forward logic in every RPC method implementation, 
```
	isLeader, needForward := s.isLeaderAndNeedForward()
	if !isLeader {
		if needForward {
			return s.leaderClient.GetTaskCfg(ctx, req) // if failed because of leader switch, should retry imediately 
		}
		return nil, terror.ErrMasterRequestIsNotForwardToLeader
	}
```
because `req` has different type (`*pb.GetTaskCfgRequest`, `*pb.OperateSchemaRequest`...), and different type has different RPC (`GetTaskCfg`, `OperateSchema`) I found it's hard to bypass Go's type system for extracting some shared logic (logging, check leader-forward...)

**Describe the feature you'd like:**

change protobuf messages (XXXRequest, XXXResponse) to a unified message, for example,
```
message CommonResponse {
    string type = 1;
    string args = 2;
}
```
using unmarshal to decode `args` into structures in current messages, and `type` used by `switch` to dispatch different RPC method when finally called.

But we need edit more place when adding RPC, and keep `type` a valid type.

**Describe alternatives you've considered:**

use `interface{}`?

**Teachability, Documentation, Adoption, Migration Strategy:**

### PR DESCRIPTION
**Pull request by lance6716**
### What problem does this PR solve? 
extract common logic to a function, if in future we will change some logic in this function it would be easier.

close #811 

### What is changed and how it works?
use `runtime` and `reflect` to achieve this

### Check List 

Tests 

 - pass original test

Code changes

Side effects

 - Possible performance regression

Related changes

 - Need to cherry-pick to the release branch

### CODE REVIEW

File: dm/master/server.go [hunk 1]
```diff
@@ -2191,3 +2171,47 @@ func (s *Server) HandleError(ctx context.Context, req *pb.HandleErrorRequest) (*
...
+// sharedLogic does some shared logic for each RPC implementation
+// arguments with `Pointer` suffix should be pointer to that variable its name indicated
+// return `true` means caller should return with variable that `xxPointer` modified
+func (s *Server) sharedLogic(ctx context.Context, req interface{}, respPointer interface{}, errPointer *error) bool {
+	pc, _, _, _ := runtime.Caller(1)
+	fullMethodName := runtime.FuncForPC(pc).Name()
+	methodName := fullMethodName[strings.LastIndexByte(fullMethodName, '.')+1:]
+
+	// origin code:
+	//	log.L().Info("", zap.Stringer("payload", req), zap.String("request", "ListMember"))
```
Comment (csuzhangxc) [inline dm/master/server.go:2184]: remove this to be clearer?

File: dm/master/server.go [hunk 2]
```diff
@@ -2191,3 +2171,47 @@ func (s *Server) HandleError(ctx context.Context, req *pb.HandleErrorRequest) (*
...
+	//	if !isLeader {
+	//		if needForward {
+	//			return s.leaderClient.ListMember(ctx, req)
+	//		}
+	//		return nil, terror.ErrMasterRequestIsNotForwardToLeader
+	//	}
+	isLeader, needForward := s.isLeaderAndNeedForward()
+	if !isLeader {
+		if needForward {
+			params := []reflect.Value{reflect.ValueOf(ctx), reflect.ValueOf(req)}
```
Comment (csuzhangxc) [inline dm/master/server.go:2198]: How about adding a log to indicate we are forwarding the request from XXX to YYY?

File: dm/master/server.go [hunk 3]
```diff
@@ -2191,3 +2171,47 @@ func (s *Server) HandleError(ctx context.Context, req *pb.HandleErrorRequest) (*
...
+	// origin code:
+	//  isLeader, needForward := s.isLeaderAndNeedForward()
+	//	if !isLeader {
+	//		if needForward {
+	//			return s.leaderClient.ListMember(ctx, req)
+	//		}
+	//		return nil, terror.ErrMasterRequestIsNotForwardToLeader
+	//	}
+	isLeader, needForward := s.isLeaderAndNeedForward()
+	if !isLeader {
```
Comment (csuzhangxc) [inline dm/master/server.go:2196]: can check `if isLeader` and `return false` to decrease indentation.

File: dm/pb/hide_password.go [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+var (
+	// HidePwdFunc should be overwrite by utils.HidePassword, this variable is for avoiding cycle import
+	HidePwdFunc = func(s string) string {
+		return s
+	}
+)
+
+// MarshalLogObject implements zapcore.ObjectMarshaler
+func (m *StartTaskRequest) MarshalLogObject(enc zapcore.ObjectEncoder) error {
+	enc.AddString("HidePasswordObject", HidePwdFunc(m.String()))
```
Comment (csuzhangxc) [inline dm/pb/hide_password.go:16]: 👍

**Review by csuzhangxc — COMMENTED** [review]
rest LGTM

**Review by csuzhangxc — COMMENTED** [review]
LGTM

**Review by GMHDBJD — APPROVED** [review]
LGTM

**Pull request merged**